### PR TITLE
JWT authentication for the router

### PR DESCRIPTION
**Pull request by garypen**
This is based on the draft design that I wrote up this morning. It's still to be refined based on feedback that I receive.

fixes: #912

Tasks:
 - [x] Configuration (to be refined)
 - [x] JWKS lazy, deduplicated retrieval (net or disk)
 - [x] Header parsing
 - [x] Header name/prefix customization
 - [x] JWKS cooldown logic
 - [x] JWKS cache clearing
 - [x] key retention
 - [x] Validation
 - [x] Claim Validation
 - [x] Claim propagation via context
 - [x] Add a rhai example interacting with claims
 - [x] testing
 - [x] documentation

### CODE REVIEW

File: apollo-router/src/plugins/authentication.rs [hunk 1]
```diff
@@ -0,0 +1,901 @@
...
+use crate::register_plugin;
+use crate::services::apollo_graph_reference;
+use crate::services::router;
+use crate::Context;
+
+type SharedDeduplicate = Arc<
+    Deduplicate<Box<dyn Fn(Url) -> DeduplicateFuture<JwkSet> + Send + Sync + 'static>, Url, JwkSet>,
+>;
+
+pub(crate) const AUTHENTICATION_SPAN_NAME: &str = "authentication plugin";
```
Comment (bnjjj) [inline apollo-router/src/plugins/authentication.rs:45]: I don't think we have spaces in our spans currently, I think it's not really an issue but maybe we might want to be consistent ? IDK

File: apollo-router/src/plugins/authentication.rs [hunk 2]
```diff
@@ -0,0 +1,901 @@
...
+}
+
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema)]
+struct JWTConf {
+    /// Retrieve our JWK Set from here
+    jwks_url: String,
+    /// HTTP header expected to contain JWT
+    #[serde(default = "default_header_name")]
+    header_name: String,
+    /// Header prefix
```
Comment (bnjjj) [inline apollo-router/src/plugins/authentication.rs:71]:
```suggestion
    /// Header value prefix
```

File: apollo-router/src/plugins/authentication.rs [hunk 3]
```diff
@@ -0,0 +1,901 @@
...
+    #[serde(default = "default_header_prefix")]
+    header_prefix: String,
```
Comment (bnjjj) [inline apollo-router/src/plugins/authentication.rs:73]:
```suggestion
    header_value_prefix: String,
```
to avoid confusion with the header name ?

File: apollo-router/src/plugins/authentication.rs [hunk 4]
```diff
@@ -0,0 +1,901 @@
...
+                            .ok_or(LicenseError::MissingGraphReference)
+                            .ok()?;
+                        let path = url.to_file_path().ok()?;
+                        read_to_string(path).await.ok()?
+                    } else {
+                        let my_client = CLIENT.as_ref().map_err(|e| e.to_string()).ok()?.clone();
+
+                        my_client
+                            .get(url)
+                            .header(ACCEPT, "application/json")
```
Comment (bnjjj) [inline apollo-router/src/plugins/authentication.rs:127]: We have a constant for this `"application/json"`

File: apollo-router/src/plugins/authentication.rs [hunk 5]
```diff
@@ -0,0 +1,901 @@
...
+                            .header(CONTENT_TYPE, "application/json")
```
Comment (bnjjj) [inline apollo-router/src/plugins/authentication.rs:128]: Same

File: apollo-router/src/plugins/authentication.rs [hunk 6]
```diff
@@ -0,0 +1,901 @@
...
+                    // Let's trim out leading and trailing whitespace to be accommodating
+                    let jwt_value = jwt_value_untrimmed.trim();
+
+                    // Make sure the format of our message matches our expectations
+                    // Technically, the spec is case sensitive, but let's accept
+                    // case variations
+                    if !jwt_value
+                        .to_uppercase()
+                        .as_str()
+                        .starts_with(&my_config.header_prefix.to_uppercase())
```
Comment (bnjjj) [inline apollo-router/src/plugins/authentication.rs:236]: Should we strip the header_prefix ?

File: apollo-router/src/plugins/authentication.rs [hunk 7]
```diff
@@ -0,0 +1,901 @@
...
+                            ) {
+                                Ok(v) => v,
+                                Err(e) => {
+                                    return failure_message(
+                                        request.context,
+                                        format!("Could not create decode JWT: {}", e),
+                                        StatusCode::UNAUTHORIZED,
+                                    );
+                                }
+                            };
```
Comment (bnjjj) [inline apollo-router/src/plugins/authentication.rs:381]: You have a lot of this pattern in this file, maybe we can create a macro to handle Option and Result ?

File: apollo-router/src/plugins/authentication.rs (hunk 4 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: done: 660f6552d824930ee265404b4d90b3221dcb6461

File: apollo-router/src/plugins/authentication.rs (hunk 5 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: also done: 660f6552d824930ee265404b4d90b3221dcb6461

File: apollo-router/src/plugins/authentication.rs (hunk 3 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: done: b8c48ba5cd7bdb418b3b0834ecf333c41479de4c

File: apollo-router/src/plugins/authentication.rs (hunk 2 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: done: b8c48ba5cd7bdb418b3b0834ecf333c41479de4c

File: apollo-router/src/plugins/authentication.rs (hunk 1 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: done: b8c48ba5cd7bdb418b3b0834ecf333c41479de4c

I also fixed up some other spans that I knew had spaces in them. I haven't checked the whole router through...

File: apollo-router/src/plugins/authentication.rs (hunk 6 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: I'm not sure what you intend. This is a case insensitive comparison of the header prefix with the expected configured value. How would stripping the `header_value_prefix` help with that?

File: apollo-router/src/plugins/authentication.rs (hunk 7 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs:424]: I could do. It's a good suggestion. I'll do it if I find some spare time. Otherwise, maybe a follow-up.

File: apollo-router/src/plugins/authentication.rs (hunk 6 above)
Comment (bnjjj) [reply apollo-router/src/plugins/authentication.rs]: It was more to avoid errors like the user has written 1 or 2 spaces at the end of the prefix in the yaml configuration and doesn't understand why it's not working. But maybe we should not think about that and don't touch a this prefix at all.

**Review by bnjjj — APPROVED** [review]

File: apollo-router/src/plugins/authentication.rs [hunk 8]
```diff
@@ -0,0 +1,944 @@
...
+                    if jwt_parts.len() != 2 {
+                        return failure_message(
+                            request.context,
+                            format!("Header Value: '{jwt_value}' is not correctly formatted. Missing JWT"),
+                            StatusCode::BAD_REQUEST,
+                        );
+                    }
+
+                    // Trim off any trailing white space (not valid in BASE64 encoding)
+                    let jwt = jwt_parts[1].trim_end();
```
Comment (Geal) [inline apollo-router/src/plugins/authentication.rs:302]: this was done already by the call to `trim` on line 271

File: apollo-router/src/plugins/authentication.rs [hunk 9]
```diff
@@ -0,0 +1,944 @@
...
+                                    request.context,
+                                    format!("Could not find kid: '{kid}' in JWKS set"),
+                                    StatusCode::UNAUTHORIZED,
+                                )
+                            }
+                        }
+                    }
+                }
+            })
+            .buffer(20_000)
```
Comment (Geal) [inline apollo-router/src/plugins/authentication.rs:488]: so, I have to ask, why is this buffered? :smile:

File: examples/jwt-claims/rhai/src/jwt_claims.rhai [hunk 10]
```diff
@@ -0,0 +1,50 @@
+// Call map_request with our service and pass in a string with the name
```
Comment (Geal) [inline examples/jwt-claims/rhai/src/jwt_claims.rhai:1]: this is a great example :+1:

File: apollo-router/src/plugins/authentication.rs [hunk 11]
```diff
@@ -0,0 +1,944 @@
...
+
+        ServiceBuilder::new()
+            .instrument(authentication_service_span())
+            .checkpoint_async(move |request: router::Request| {
+                let my_config = request_full_config.clone();
+                let my_jwks = request_jwks.clone();
+                let my_jwks_url = request_jwks_url.clone();
+                const AUTHENTICATION_KIND: &str = "JWT";
+
+                async move {
```
Comment (Geal) [inline apollo-router/src/plugins/authentication.rs:217]: could this be extracted to a function?

File: apollo-router/src/plugins/authentication.rs [hunk 12]
```diff
@@ -0,0 +1,944 @@
...
+                                "Could not find JWKS set at the configured location".to_string(),
+                                StatusCode::INTERNAL_SERVER_ERROR,
+                            );
+                        }
+                    };
+
+                    // Now let's try to validate our token
+                    match jwks.find(&kid) {
+                        Some(jwk) => {
+                            let decoding_key = match DecodingKey::from_jwk(jwk) {
```
Comment (Geal) [inline apollo-router/src/plugins/authentication.rs:387]: could you extract the code to download the JWKS, deduplicate the call and handling the cooldown in one function like `find_jwks(jwks: Arc<Deduplicate<?>>, kid: &str) -> Result<DecodingKey, String>`?
It is a bit hard to follow here, with de JWKS downloading code, then JWT validation in the middle, then again JWKS download and cooldown code.

File: apollo-router/src/plugins/authentication.rs [hunk 13]
```diff
@@ -0,0 +1,944 @@
...
+                    //  - we can't retrieve a JWKS (a)
+                    //  - the retrieved JWKS is None (b)
+                    //  - we have a kid that we don't know about (c)
+                    // We need to do some additional processing.  (I've tagged with comments a/b/c below)
+                    let err_cleanup = move |cache_clear| {
+                        // If cache_clear is set, then we clear the cache so that subsequent
+                        // requests will retrieve new JWKS.
+                        // The COOLDOWN controls attempts to retrieve based on a new "kid", but not
+                        // repeated attempts due to failure to retrieve JWKS.
+                        if cache_clear {
```
Comment (Geal) [inline apollo-router/src/plugins/authentication.rs:340]: why is `cache_clear` used here? Was there a previous version with an "else" branch? Because here `cache_clear` could be removed and we can replace `err_cleanup(true)` with `err_cleanup()`, and remove the calls to `err_cleanup(false)` at lines 363 (a) and 375 (b).
It also feels like `err_cleanup` might not be the right name for a rate limiter

**Review by Geal — CHANGES_REQUESTED** [review]

File: apollo-router/src/plugins/authentication.rs (hunk 7 above)
Comment (Geal) [reply apollo-router/src/plugins/authentication.rs:424]: another approach here would be to make it as a function that returns a `Result<serde_json::Value, (String, StatusCode)>` and then have a wrapper that calls that function, and calls `failure_message` on the err case

File: apollo-router/src/plugins/authentication.rs [hunk 14]
```diff
@@ -0,0 +1,944 @@
...
+#[async_trait::async_trait]
+impl Plugin for AuthenticationPlugin {
+    type Config = Conf;
+
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        let url: Url = Url::from_str(&init.config.experimental.jwt.jwks_url)?;
+        // This function is expected to return an Optional value, but we'd like to let
+        // users know the various failure conditions. Hence the various clumsy map_err()
+        // scattered through the processing.
+        let getter: Box<dyn Fn(Url) -> DeduplicateFuture<JwkSet> + Send + Sync + 'static> =
```
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs:119]: I don't mind using `DeduplicateFuture` since it's the type that the crate expects and saves me from forgetting to specify the future contains an `Option`. I think it requires more thought to use `BoxFuture`.

I also don't mind having closures that don't capture anything from their environment. I know it could be a function, but I think it's just as readable to have a closure here.

I could change both of these, but it does feel as though the changes are purely stylistic.

File: apollo-router/src/plugins/authentication.rs [hunk 15]
```diff
@@ -0,0 +1,944 @@
...
+                            format!(
+                                "Header Value: '{jwt_value_untrimmed}' is not correctly formatted. prefix should be '{}'",
+                                my_config.header_value_prefix
+                            ),
+                            StatusCode::BAD_REQUEST,
+                        );
+                    }
+
+                    // Split our string in (at most 2) sections.
+                    let jwt_parts: Vec<&str> = jwt_value.splitn(2, ' ').collect();
```
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs:292]: The documentation does say: [don't do that](https://github.com/apollographql/router/blob/b8c48ba5cd7bdb418b3b0834ecf333c41479de4c/docs/source/configuration/authn-jwt.mdx). I'll add a check to the `Plugin::new` code to make sure there are no "blanks" in the `header_value_prefix`.

File: apollo-router/src/plugins/authentication.rs (hunk 9 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs:488]: If I don't buffer the service, then I get:

```
error[E0277]: the trait bound `tower::util::BoxService<services::router::Request, services::router::Response, Box<(dyn StdError + std::marker::Send + std::marker::Sync + 'static)>>: Clone` is not satisfied
   --> apollo-router/src/plugins/authentication.rs:501:22
    |
501 |             .service(service)
    |              ------- ^^^^^^^ the trait `Clone` is not implemented for `tower::util::BoxService<services::router::Request, services::router::Response, Box<(dyn StdError + std::marker::Send + std::marker::Sync + 'static)>>`
    |              |
    |              required by a bound introduced by this call
    |
    = help: the trait `tower::Layer<S>` is implemented for `Stack<Inner, Outer>`
<plus about another couple of hundred lines of errror...>
```

This all stems from `checkpoint_async` which requires the service to be `Clone`. I presume you are really asking "why not make it a layer rather than a plugin?".

File: apollo-router/src/plugins/authentication.rs (hunk 13 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs:340]: Good catch. You are right, there was a branch here originally. I'll clean this up as you suggest.

Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: done: 1bd8ed69f1d810adef32d3ad57d20d02328c0f6d

File: apollo-router/src/plugins/authentication.rs (hunk 8 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: removed: 1bd8ed69f1d810adef32d3ad57d20d02328c0f6d

File: apollo-router/src/plugins/authentication.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs:304]: added check for whitespace as decribed and added two new tests: 1bd8ed69f1d810adef32d3ad57d20d02328c0f6d

File: apollo-router/src/plugins/authentication.rs [hunk 16]
```diff
@@ -0,0 +1,944 @@
...
+                    // Let's trim out leading and trailing whitespace to be accommodating
+                    let jwt_value = jwt_value_untrimmed.trim();
+
+                    // Make sure the format of our message matches our expectations
+                    // Technically, the spec is case sensitive, but let's accept
+                    // case variations
+                    if !jwt_value
+                        .to_uppercase()
+                        .as_str()
+                        .starts_with(&my_config.header_value_prefix.to_uppercase())
```
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: done: 1bd8ed69f1d810adef32d3ad57d20d02328c0f6d

File: apollo-router/src/plugins/authentication.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/plugins/authentication.rs:130]: yes, this is purely stylistic (although what I propose would avoid one negligible `Box`), but this is just to transmit the general feeling of "I don't understand what's going on" when reading that code, and proposing ways to make it more obvious

File: apollo-router/src/plugins/authentication.rs (hunk 9 above)
Comment (Geal) [reply apollo-router/src/plugins/authentication.rs]: this can land with the buffer in there, we'll see if we need to remove it later

File: apollo-router/src/plugins/authentication.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (Geal) [reply apollo-router/src/plugins/authentication.rs:304]: a "don't do that" in doc works better with an enforced check :+1:

File: apollo-router/src/plugins/authentication.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs:130]: Sure. I want the code to be readable, but I guess I was finding it readable already. I'll make the change you suggest and separate it out from `Plugin::new()`

File: apollo-router/src/plugins/authentication.rs (hunk 11 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs:233]: probably, but most of the other plugins using checkpoint_async are just written to use an async block.

File: apollo-router/src/plugins/authentication.rs (hunk 12 above)
Comment (garypen) [reply apollo-router/src/plugins/authentication.rs]: I've re-factored the COOLDOWN code to move it to remove the closure and place the functionality in the only place where it's used now (unrecognised kid): 66935caf895ea2db237e9586dcb78866e6c956c8

So: I think this is much clearer and I don't think the proposed `find_jwks` is required. What do you think?

Comment (Geal) [reply apollo-router/src/plugins/authentication.rs]: this is way easier to follow, thanks

**Review by Geal — APPROVED** [review]
